### DISCUSSION TITLE
Delete multiple nodes from document by ID

### DISCUSSION THREAD

**Discussion by dkrym** [ueberdosis/tiptap#2252] [General Questions] [answered]
Hi everybody,

I have a custom block node with uuid and I want to delete several of them in document. What would be the best way to do it? Is it possible to do it in one pass? I am using `doc.descendants` to get the position but the obvious issue is I need to remove multiple nodes from document so I assume the `pos` in `descendants` function is incorrect after first deleting.

Is it possible to use `doc.descendants((node, pos) => { ... })` for deleting multiple nodes and keep the `pos` correct?

I came up with this which looks terrible (simplified version), but it seems it works... but there has to be a better way

```javascript
     let ch = this.editor.chain();

     ids.forEach((id) => {
        ch = ch.command(({ tr, commands }) => {
          tr.doc.descendants((node, pos) => {
            if (node.type.name == 'custom_item') {
              if (id === node.attrs.uuid) {
                commands.deleteRange({
                  from: pos,
                  to: pos + node.nodeSize,
                });
              }
              return false;
            }
          });

          return true;
        });
      });

      ch.run();
```

**Comment by philippkuehn** [comment, marked as answer]
There is a [`forEach`](https://tiptap.dev/api/commands/for-each) command and a `findChildren` helper which could make this more readable. This code should create one single transaction:

```js
import { findChildren } from '@tiptap/core'

editor.commands.forEach(ids, (id, { tr, commands }) => {
  const item = findChildren(tr.doc, node => {
    return node.type.name === 'custom_item' && id === node.attrs.uuid
  })?.[0]

  if (!item) {
    return true
  }

  return commands.deleteRange({
    from: item.pos,
    to: item.pos + item.node.nodeSize,
  })
})
```

**Comment by dkrym** [comment]
cheers, I will try this out
